File: Mac/HadesApp/Sources/HadesApp/MainWindow/MemoryDocumentFetchState.swift
import HadesControl

/// One selected document's fetch state for the Memory view - `GET /control/memory/document`,
/// resolved. `MemoryViewModel.selectDocument(name:)` is the only thing that changes this; nothing
/// polls it on a timer - the same shape `TraceDetailFetchState` already established for a selected
/// trace's span detail: a document the user opened to read/edit is a fixed snapshot for as long as
/// they are looking at it, not a live value `refresh()` should silently overwrite out from under an
/// in-progress edit.
public enum MemoryDocumentFetchState: Equatable, Sendable {
    /// No document selected yet - the ordinary starting state, not an error.
    case notSelected

    /// The full raw content for the most recently selected document.
    case loaded(MemoryDocumentResult)

    /// `GET /control/memory/document` answered with a server error (e.g. `'{name}' does not exist
    /// yet.`) - `message` is `ControlClientError.server(message:)`'s payload, verbatim.
    case failed(message: String)
}

File: Windows/Hades.Shell/ViewModels/MemoryDocumentFetchState.cs
using Hades.Control.Client.Dtos;

namespace Hades.Shell.ViewModels;

public enum MemoryDocumentFetchKind
{
    /// <summary>No document selected yet - the ordinary starting state, not an error.</summary>
    NotSelected,

    /// <summary>The full raw content for the most recently selected document.</summary>
    Loaded,

    /// <summary>The fetch answered with a server error; the message is the server's own.</summary>
    Failed,
}

/// <summary>
/// One selected document's fetch state. The port of
/// <c>Mac/HadesApp/Sources/HadesApp/MainWindow/MemoryDocumentFetchState.swift</c>, and the same shape
/// <see cref="TraceDetailFetchState"/> already established.
///
/// Only <see cref="MemoryViewModel.SelectDocumentAsync"/> changes this, and nothing polls it: a
/// document opened to read or edit is a fixed snapshot for as long as someone is looking at it, not
/// a live value a tick should silently overwrite out from under an in-progress edit.
/// </summary>
public readonly record struct MemoryDocumentFetchState
{
    public MemoryDocumentFetchKind Kind { get; }

    /// <summary>Set when <see cref="Kind"/> is Loaded.</summary>
    public MemoryDocumentResult? Document { get; }

    /// <summary>Set when <see cref="Kind"/> is Failed - e.g. "'{name}' does not exist yet.",
    /// verbatim.</summary>
    public string? Message { get; }

    MemoryDocumentFetchState(MemoryDocumentFetchKind kind, MemoryDocumentResult? document, string? message)
    {
        Kind = kind;
        Document = document;
        Message = message;
    }

    public static readonly MemoryDocumentFetchState NotSelected =
        new(MemoryDocumentFetchKind.NotSelected, null, null);

    public static MemoryDocumentFetchState Loaded(MemoryDocumentResult document) =>
        new(MemoryDocumentFetchKind.Loaded, document, null);

    public static MemoryDocumentFetchState Failed(string message) =>
        new(MemoryDocumentFetchKind.Failed, null, message);
}
